### PR TITLE
Improve ractor safety

### PR DESCRIPTION
**Pull request by HoneyryderChuck**
Also, replacing usage of refinements.

### CODE REVIEW

File: lib/http/2/header/huffman_statemachine.rb [hunk 1]
```diff
@@ -268,7 +268,7 @@ class Huffman
         [[28, 29], [28, 5], [29, 29], [29, 5], [30, 29], [30, 5], [31, 29], [31, 5], [127, 29], [127, 5], [220, 29], [220, 5], [249, 29], [249, 5], [nil, 254], [nil, 255]],
         [[10, 17], [10, 18], [10, 19], [10, 20], [10, 21], [10, 22], [10, 23], [10, 7], [13, 17], [13, 18], [13, 19], [13, 20], [13, 21], [13, 22], [13, 23], [13, 7]],
         [[22, 17], [22, 18], [22, 19], [22, 20], [22, 21], [22, 22], [22, 23], [22, 7], [256, 17], [256, 18], [256, 19], [256, 20], [256, 21], [256, 22], [256, 23], [256, 7]],
-      ].each { |arr| arr.each { |subarr| subarr.each(&:freeze) }.freeze }.freeze
+      ].each { |arr| arr.each { |subarr| subarr.freeze }.freeze }.freeze
```
Comment (mullermp) [inline lib/http/2/header/huffman_statemachine.rb:271]: Nice. This looks like it was a bug right? Only the arrays need to be frozen not values?

File: lib/http/2/header/encoding_context.rb [hunk 2]
```diff
@@ -77,10 +77,11 @@ class EncodingContext
...
 
-      STATIC_TABLE_BY_FIELD = STATIC_TABLE
-                              .each_with_object({})
-                              .with_index { |((field, value), hs), idx| (hs[field] ||= []) << [idx, value] }
-                              .each { |pair| pair.each(&:freeze).freeze }.freeze
+      STATIC_TABLE_BY_FIELD =
+        STATIC_TABLE
+        .each_with_object({})
+        .with_index { |((field, value), hs), idx| (hs[field.freeze] ||= []) << [idx, value.freeze].freeze }
+        .each { |pair| pair.each(&:freeze).freeze }.freeze
```
Comment (mullermp) [inline lib/http/2/header/encoding_context.rb:84]: Same thing here right? Do pair values need to be frozen or only the arrays? There are probably instances of this in other places.

File: lib/http/2/header/encoding_context.rb [hunk 3]
```diff
@@ -77,10 +77,11 @@ class EncodingContext
...
       ].each { |pair| pair.each(&:freeze).freeze }.freeze
 
-      STATIC_TABLE_BY_FIELD = STATIC_TABLE
-                              .each_with_object({})
-                              .with_index { |((field, value), hs), idx| (hs[field] ||= []) << [idx, value] }
-                              .each { |pair| pair.each(&:freeze).freeze }.freeze
+      STATIC_TABLE_BY_FIELD =
+        STATIC_TABLE
+        .each_with_object({})
+        .with_index { |((field, value), hs), idx| (hs[field.freeze] ||= []) << [idx, value.freeze].freeze }
```
Comment (mullermp) [inline lib/http/2/header/encoding_context.rb:83]: Ditto on freeze question

**Review by mullermp — APPROVED** [review]
Nice clean up. Just added some considerations, it's otherwise fine.

File: lib/http/2/header/huffman_statemachine.rb (hunk 1 above)
Comment (HoneyryderChuck) [reply lib/http/2/header/huffman_statemachine.rb:271]: exactly. not a "bug" as in, it didn't lead to breaking code, but it made this constant unusable across ractors.

File: lib/http/2/header/encoding_context.rb (hunk 2 above)
Comment (HoneyryderChuck) [reply lib/http/2/header/encoding_context.rb:84]: indeed, arrays must be frozen to be shareable (values already were).

File: lib/http/2/header/encoding_context.rb (hunk 3 above)
Comment (HoneyryderChuck) [reply lib/http/2/header/encoding_context.rb:83]: same

File: lib/http/2/header/encoding_context.rb (hunk 2 above)
Comment (mullermp) [reply lib/http/2/header/encoding_context.rb:84]: I'm essentially asking if you need `pair.each(&:freeze)` here? I would assume you don't want values frozen.

File: lib/http/2/header/encoding_context.rb (hunk 3 above)
Comment (mullermp) [reply lib/http/2/header/encoding_context.rb:83]: Do you need `value.freeze`? I thought just arrays needed to be.

File: lib/http/2/header/encoding_context.rb (hunk 2 above)
Comment (HoneyryderChuck) [reply lib/http/2/header/encoding_context.rb]: you're right, I've simplified it now, and not try to freeze values that I know are already frozen.

**Pull request merged**